pinot-query-runtime/src/main/java/org/apache/pinot/query/mailbox/ReceivingMailbox.java

```diff
@@ -51,16 +52,30 @@ public class ReceivingMailbox {
   // TODO: Revisit if this is the correct way to apply back pressure
   private final BlockingQueue<TransferableBlock> _blocks = new ArrayBlockingQueue<>(DEFAULT_MAX_PENDING_BLOCKS);
   private final AtomicReference<TransferableBlock> _errorBlock = new AtomicReference<>();
+  @Nullable
+  private volatile Reader _reader;
```

Comment:
if we have the reader API. do we still need the callback registration? (we can remove if this is fully covered by the reader callback

Reply:
discussed offline. let's keep the callback for now b/c 
1. PB uses it to indicate whether all PB worker finishes (which can be optimized). and 
2. callback is on a OpChain level and reader is on a mailbox level thus it is not a "superset" of functionality